### ISSUE TITLE
[加密货币泡沫检测] | Crypto Bubble Detection

### ISSUE THREAD

**Issue by Kaoruha** [Kaoruha/GinkGO#4462]
**标签**: quantitative-strategy
### 摘要
使用改进的LPPL模型和情绪指标检测加密货币泡沫，实现泡沫阶段的择时做空策略。
### 算法描述
**核心原理**：
加密货币泡沫表现为超指数增长，价格服从Log-Periodic Power Law (LPPL)模型，通过拟合模型参数识别泡沫阶段。

**LPPL模型**：
ln(p(t)) = A + B(tc - t)^m + C(tc - t)^m * cos(ω*ln(tc - t) - φ)

其中：
- tc: 临界时间（泡沫破裂点）
- m: 泡沫增长率（0 < m < 1）
- ω: 对数周期振荡频率
- φ: 相位

**关键参数解释**：
1. **泡沫指数**：tc - t < 20天 且 m > 0.4
2. **崩盘概率**：基于LPPL拟合误差和资金流向
3. **RSI背离**：价格创新高但RSI下降

**交易逻辑**：
- Phase 1 (0 < tc-t < 60): 观察期，不交易
- Phase 2 (20 < tc-t < 40): 轻仓做空
- Phase 3 (tc-t < 20): 重仓做空

**风控要点**：
- 严格止损（5%）
- 分批建仓
- 结合交易所情绪数据
### Python代码示例
```python
import numpy as np
from scipy.optimize import minimize, curve_fit
from scipy.stats import norm
from datetime import datetime, timedelta
from typing import Dict, List, Tuple, Optional
from dataclasses import dataclass
from enum import Enum

@dataclass
class BubbleSignal:
    timestamp: datetime
    price: float
    tc_estimate: datetime  # 估计的临界时间
    bubble_index: float   # 泡沫指数 (0-1)
   崩盘概率: float
    signal: int  # -1=做空, 0=观望, 1=做多
    confidence: float

class LPPLModel:
    """Log-Periodic Power Law模型"""
    
    def __init__(self):
        self.params = None
        self.residuals = None
    
    @staticmethod
    def lppl_func(t: np.ndarray, A: float, B: float, C: float, 
                  tc: float, m: float, omega: float, phi: float) -> np.ndarray:
        """LPPL函数"""
        dt = np.maximum(tc - t, 1e-10)
        return A + B * np.power(dt, m) + C * np.power(dt, m) * np.cos(
            omega * np.log(dt) - phi
        )
    
    def fit(self, prices: np.ndarray, dates: np.ndarray) -> Dict:
        """拟合LPPL模型"""
        t = np.arange(len(prices))
        
        # 初始化参数
        p0 = [
            np.log(prices[-1]),      # A
            -0.1,                     # B (负值表示上涨)
            0.1,                      # C
            len(prices) + 30,         # tc
            0.5,                      # m
            8.0,                      # omega
            np.pi / 6                 # phi
        ]
        
        bounds = (
            [np.log(prices.min()), -1, -1, len(prices), 0.01, 4, 0],
            [np.log(prices.max()) * 1.5, 0, 1, len(prices) + 120, 0.9, 15, 2*np.pi]
        )
        
        try:
            popt, pcov = curve_fit(
                self.lppl_func, t, np.log(prices),
                p0=p0, bounds=bounds, maxfev=10000
            )
            self.params = popt
            
            # 计算拟合误差
            predicted = self.lppl_func(t, *popt)
            self.residuals = np.log(prices) - predicted
            
            return {
                "A": popt[0], "B": popt[1], "C": popt[2],
                "tc": popt[3], "m": popt[4], "omega": popt[5], "phi": popt[6],
                "r_squared": 1 - np.var(self.residuals) / np.var(np.log(prices))
            }
        except Exception as e:
            return None
    
    def predict_tc(self) -> Optional[float]:
        """预测临界时间"""
        if self.params is None:
            return None
        return self.params[3]

@dataclass
class SentimentData:
    fear_greed_index: float      # 0-100
    funding_rate: float          # 资金费率
    open_interest: float         # 未平仓量
    volume_24h: float            # 24h成交量
    whale_ratio: float           # 鲸鱼占比

class CryptoBubbleDetector:
    """加密货币泡沫检测器"""
    
    def __init__(
        self,
        min_price_history: int = 60,
        bubble_threshold: float = 0.5,
        crash_prob_threshold: float = 0.6,
        m_upper: float = 0.6,
        m_lower: float = 0.3
    ):
        self.min_history = min_price_history
        self.bubble_threshold = bubble_threshold
        self.crash_threshold = crash_prob_threshold
        self.m_upper = m_upper
        self.m_lower = m_lower
        self.lppl_model = LPPLModel()
        self.price_history: List[float] = []
        self.signals: List[BubbleSignal] = []
    
    def calculate_rsi(self, prices: np.ndarray, period: int = 14) -> float:
        """计算RSI"""
        if len(prices) < period + 1:
            return 50.0
        
        deltas = np.diff(prices)
        gains = np.where(deltas > 0, deltas, 0)
        losses = np.where(deltas < 0, -deltas, 0)
        
        avg_gain = np.mean(gains[-period:])
        avg_loss = np.mean(losses[-period:])
        
        if avg_loss < 1e-10:
            return 100.0
        
        rs = avg_gain / avg_loss
        return 100 - 100 / (1 + rs)
    
    def calculate_hurst_exponent(self, prices: np.ndarray) -> float:
        """计算Hurst指数"""
        if len(prices) < 20:
            return 0.5
        
        lags = range(2, min(20, len(prices) // 2))
        tau = [np.std(np.subtract(prices[lag:], prices[:-lag])) for lag in lags]
        
        poly = np.polyfit(np.log(list(lags)), np.log(tau), 1)
        return poly[0]
    
    def detect_bubble(
        self,
        price: float,
        timestamp: datetime,
        sentiment: Optional[SentimentData] = None
    ) -> BubbleSignal:
        """检测泡沫"""
        self.price_history.append(price)
        
        if len(self.price_history) < self.min_history:
            return BubbleSignal(
                timestamp=timestamp,
                price=price,
                tc_estimate=timestamp + timedelta(days=30),
                bubble_index=0.0,
                crash_prob=0.0,
                signal=0,
                confidence=0.0
            )
        
        prices = np.array(self.price_history)
        
        # 拟合LPPL模型
        fit_result = self.lppl_model.fit(prices, np.arange(len(prices)))
        
        if fit_result is None:
            return BubbleSignal(
                timestamp=timestamp,
                price=price,
                tc_estimate=timestamp + timedelta(days=30),
                bubble_index=0.0,
                crash_prob=0.0,
                signal=0,
                confidence=0.0
            )
        
        # 计算泡沫指数
        m = fit_result["m"]
        t_current = len(prices)
        tc_estimated = fit_result["tc"]
        days_to_critical = tc_estimated - t_current
        
        # 泡沫指数：越接近tc，m越大，泡沫越明显
        if days_to_critical > 0:
            bubble_index = max(0, 1 - days_to_critical / 60)
            bubble_index *= (m / 0.7)  # 调整m的影响
        else:
            bubble_index = 1.0
        
        # RSI检测
        rsi = self.calculate_rsi(prices)
        rsi_divergence = 0.0
        if len(prices) > 20:
            if prices[-1] > prices[-20] and rsi < 60:
                rsi_divergence = 0.5  # 价格创新高但RSI不跟随
        
        # 崩盘概率
        crash_prob = 0.0
        if fit_result["r_squared"] > 0.8 and bubble_index > 0.5:
            crash_prob = min(1.0, bubble_index * 0.8 + rsi_divergence * 0.2)
        
        # 生成信号
        signal = 0
        if days_to_critical < 20 and m > self.m_upper:
            signal = -1  # 重仓做空
        elif days_to_critical < 40 and m > self.m_lower:
            signal = -1  # 轻仓做空
        
        # 情绪过滤
        confidence = bubble_index
        if sentiment:
            if sentiment.fear_greed_index < 30:  # 极度恐惧不做空
                confidence *= 0.5
            if abs(sentiment.funding_rate) > 0.1:  # 高资金费率预警
                confidence *= 1.2
        
        confidence = min(1.0, confidence)
        
        return BubbleSignal(
            timestamp=timestamp,
            price=price,
            tc_estimate=timestamp + timedelta(days=max(0, int(days_to_critical))),
            bubble_index=bubble_index,
            crash_prob=crash_prob,
            signal=signal,
            confidence=confidence
        )
    
    def calculate_position_size(
        self,
        signal: BubbleSignal,
        portfolio_value: float,
        max_position: float = 0.1
    ) -> Dict:
        """计算仓位"""
        if signal.signal == 0:
            return {"size": 0, "stop_loss": 0}
        
        notional = portfolio_value * max_position * signal.confidence
        
        # 止损设置
        if signal.signal == -1:
            stop_loss = signal.price * 1.05  # 5%止损
        else:
            stop_loss = signal.price * 0.95
        
        return {
            "size": notional / signal.price,
            "stop_loss": stop_loss,
            "signal": "short" if signal.signal == -1 else "long"
        }

class CryptoBubbleStrategy:
    """加密货币泡沫策略"""
    
    def __init__(self, initial_capital: float = 100000):
        self.capital = initial_capital
        self.positions = []
        self.history = []
        self.detector = CryptoBubbleDetector()
    
    def run_backtest(
        self,
        price_data: List[Tuple[datetime, float]],
        sentiment_data: Optional[List[SentimentData]] = None
    ) -> Dict:
        """回测"""
        for i, (ts, price) in enumerate(price_data):
            sentiment = sentiment_data[i] if sentiment_data else None
            signal = self.detector.detect_bubble(price, ts, sentiment)
            
            if signal.signal != 0 and signal.confidence > 0.5:
                pos = self.detector.calculate_position_size(signal, self.capital)
                
                if signal.signal == -1:
                    # 做空：价格下跌赚钱
                    pnl = pos["size"] * (signal.price - price) if price < signal.price else 0
                else:
                    pnl = pos["size"] * (price - signal.price) if price > signal.price else 0
                
                self.capital += pnl
            
            self.history.append({
                "timestamp": ts,
                "price": price,
                "bubble_index": signal.bubble_index,
                "capital": self.capital
            })
        
        return {
            "final_capital": self.capital,
            "total_return": (self.capital - 100000) / 100000,
            "history": self.history
        }

# 测试
if __name__ == "__main__":
    np.random.seed(42)
    
    # 模拟泡沫数据：超指数增长
    t = np.arange(100)
    prices = 1000 * np.exp(0.15 * t + 0.3 * np.sin(t * 0.5))
    prices += np.random.randn(100) * prices * 0.02  # 添加噪声
    
    price_data = [(datetime.now() + timedelta(days=i), float(prices[i])) for i in range(100)]
    
    strategy = CryptoBubbleStrategy()
    result = strategy.run_backtest(price_data)
    
    print(f"Final Capital: ${result['final_capital']:,.2f}")
    print(f"Total Return: {result['total_return']:.2%}")
